streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -786,6 +822,13 @@ void runOnce() {
         // TODO: we will process some tasks even if the state is not RUNNING, i.e. some other
         // tasks are still being restored.
         if (taskManager.hasActiveRunningTasks()) {
+            // because we don't want to start a transaction if all tasks are in restoring phase,
+            // we need to track transaction state manually via `transactionInFlight`
+            if (eosBetaEnabled && !eosUpgradeModeEnabled && !transactionInFlight) {
+                producer.beginTransaction();
```

Comment:
Can we delay calling beginTxn when we are about to send / commit and found that `transactionInFlight == false`?

Reply:
Maybe -- but I am not sure what we would gain? As we commit each 100ms anyway, I don't think that hitting a timeout is a potential issue if there is at least one active task (for the restore case it's different as restoring can take longer than 10 seconds and thus we would hit transaction timeout easily).

Also we would need to add this check at multiple places in the code; atm, it's more centralized what seems to be better?